### PR TITLE
entgql: generate create/update mutation inputs

### PR DESCRIPTION
**Pull request by giautm**
Related https://github.com/ent/ent/issues/2351

### CODE REVIEW

File: entgql/internal/todo/ent/gql_mutation_input.go [hunk 1]
```diff
@@ -39,9 +39,7 @@ type CreateCategoryInput struct {
 func (i *CreateCategoryInput) Mutate(m *CategoryCreate) {
 	m.SetText(i.Text)
 	m.SetStatus(i.Status)
-	if v := i.Config; v != nil {
-		m.SetConfig(*v)
-	}
+	m.SetConfig(i.Config)
```
Comment (giautm) [inline entgql/internal/todo/ent/gql_mutation_input.go]: Updated to correct code

**Comment by crossworth** [comment]
Hello @giautm , there a few things that I think we should change on the create mutations template that I have been fixing on my local copy and would like to let you know.

- The skip annotation should not be used to skip the mutation generation, a new annotation should be created only for the mutation part, like, I like to be able to order/filter by `created_time`, `updated_time`, but I dont want to allow a mutation on it.
- How to solve the problem of a null field vs a a field not provided, is the `ClearType bool` the best solution? Maybe we should try read the provided query and handle in a different way? (to make the javascript part more easy/compatible).

**Comment by giautm** [comment]
> I like to be able to order/filter by created_time, updated_time, but I dont want to allow a mutation on it.

You can use `entgql.Skip()` on each fields you want to ignore.

**Comment by giautm** [comment]
> * I like to be able to order/filter by `created_time`, `updated_time`, but I dont want to allow a mutation

Hey @crossworth, I think the template really depends on `MutableFields`.

**Comment by giautm** [comment]
hello @crossworth now, you can use `entgql.Skip(entgql.SkipMutationCreateInput)` to skip generate Create<Type>Input!

File: entgql/schema.go [hunk 2]
```diff
@@ -478,6 +488,128 @@ func (e *schemaGenerator) buildWhereInput(t *gen.Type, gqlType string) (*ast.Def
...
+			if ant.Skip.Is(SkipMutationCreateInput) {
+				continue
+			}
+
+			if e.Unique {
+				def.Fields = append(def.Fields, &ast.FieldDefinition{
+					Name: camel(e.Name + "_ID"),
+					Type: namedType("ID", true),
+				}, &ast.FieldDefinition{
+					Name: camel(snake(e.MutationClear())),
```
Comment (giautm) [inline entgql/schema.go:593]: @a8m: I'm not sure it outputs the same result.

File: entgql/schema.go [hunk 3]
```diff
@@ -478,6 +488,128 @@ func (e *schemaGenerator) buildWhereInput(t *gen.Type, gqlType string) (*ast.Def
...
+					Type: namedType("Boolean", true),
+				})
+			} else {
+				def.Fields = append(def.Fields, &ast.FieldDefinition{
+					Name: camel("add_" + singular(e.Name) + "_IDs"),
+					Type: namedType("[ID!]", true),
+				}, &ast.FieldDefinition{
+					Name: camel("remove_" + singular(e.Name) + "_IDs"),
+					Type: namedType("[ID!]", true),
+				})
```
Comment (giautm) [inline entgql/schema.go:603]: Due to a "bug" with "IDs", I can't use `MutationAdd()` and `MutationRemove()` functions

`camel(snake(e.MutationRemove()))` return `removeGroupIds`
and `camel(e.MutationRemove())` return `removegroupids`

So, I have to change copy it logic: `camel("remove_" + singular(e.Name) + "_IDs")`

File: entgql/schema.go [hunk 4]
```diff
@@ -229,6 +231,15 @@ func (e *schemaGenerator) buildTypes(g *gen.Graph, s *ast.Schema) error {
...
 			}
 		}
+
+		if e.genMutations {
+			defs, err := e.buildMutationInputs(node, ant, gqlType)
+			if err != nil {
+				return err
+			} else if len(defs) > 0 {
+				s.AddTypes(defs...)
+			}
```
Comment (a8m) [inline entgql/schema.go:241]:
```suggestion
			defs, err := e.buildMutationInputs(node, ant, gqlType)
			if err != nil {
				return err
			}
			if if len(defs) > 0 {
				s.AddTypes(defs...)
			}
```
No need for else.

File: entgql/schema.go [hunk 5]
```diff
@@ -478,6 +489,129 @@ func (e *schemaGenerator) buildWhereInput(t *gen.Type, gqlType string) (*ast.Def
...
+			def.Fields = append(def.Fields, &ast.FieldDefinition{
+				Name:        camel(f.Name),
+				Type:        namedType(scalar, true),
+				Description: f.Comment(),
+			})
+			if f.Optional {
+				def.Fields = append(def.Fields, &ast.FieldDefinition{
+					Name: camel("clear_" + f.Name),
+					Type: namedType("Boolean", true),
+				})
```
Comment (a8m) [inline entgql/schema.go:578]:
```suggestion
				def.Fields = append(def.Fields, &ast.FieldDefinition{
					Name: "clear" + f.StructField(),
					Type: namedType("Boolean", true),
				})
```

File: entgql/schema.go (hunk 2 above)
Comment (a8m) [reply entgql/schema.go:595]: "clear" + f.StructField()

File: entgql/schema.go [hunk 6]
```diff
@@ -478,6 +489,129 @@ func (e *schemaGenerator) buildWhereInput(t *gen.Type, gqlType string) (*ast.Def
...
+			if err != nil {
+				return nil, err
+			}
+			if ant.Skip.Is(SkipMutationCreateInput) {
+				continue
+			}
+
+			if e.Unique {
+				def.Fields = append(def.Fields, &ast.FieldDefinition{
+					Name: camel(e.Name + "_ID"),
```
Comment (a8m) [inline entgql/schema.go:592]: ditto

Comment (giautm) [inline entgql/schema.go:592]:
```suggestion
					Name: camel(e.Name) + "ID",
```

File: entgql/schema.go (hunk 2 above)
Comment (giautm) [reply entgql/schema.go:595]: Screen Shot 2022-05-12 at 22 13 31
I think also need to change the go template.

File: entgql/schema.go [hunk 7]
```diff
@@ -478,6 +490,100 @@ func (e *schemaGenerator) buildWhereInput(t *gen.Type, gqlType string) (*ast.Def
...
+						Type: namedType("[ID!]", e.Optional),
+					})
+				} else {
+					def.Fields = append(def.Fields, &ast.FieldDefinition{
+						Name: camel("add_" + singular(e.Name) + "_IDs"),
+						Type: namedType("[ID!]", true),
+					}, &ast.FieldDefinition{
+						Name: camel("remove_" + singular(e.Name) + "_IDs"),
+						Type: namedType("[ID!]", true),
+					})
```
Comment (a8m) [inline entgql/schema.go:577]:
```suggestion
						Name: "add" + pascal(singular(e.Name)) + "IDs",
						Type: namedType("[ID!]", true),
					}, &ast.FieldDefinition{
						Name: "remove" + pascal(singular(e.Name)) + "IDs",
						Type: namedType("[ID!]", true),
					})
```

File: entgql/annotation.go [hunk 8]
```diff
@@ -335,6 +350,35 @@ func (a queryFieldAnnotation) Directives(directives ...Directive) queryFieldAnno
 	return a
 }
 
+type MutationOption interface {
+	IsCreate() bool
+}
+
+type builtinMutation bool
+
+func (v builtinMutation) IsCreate() bool { return bool(v) }
```
Comment (a8m) [inline entgql/annotation.go:359]: Why not use functional options that can change the `Options` object (instead of interface). Feels easier to write.

**Review by a8m — APPROVED** [review]

**Pull request merged**